### PR TITLE
🐛 amp-next-page: Fix relative URL handling when served from the cache

### LINKED ISSUE #18508: amp-next-page wrongly rewrites recommendation box URLs when relative URLS are used in the script json
**Please only file bugs/feature requests for AMP here.**

- If you have questions about how to use AMP or other general questions about AMP please ask them on Stack Overflow under the AMP HTML tag instead of filing an issue here: http://stackoverflow.com/questions/tagged/amp-html
- If you have questions/issues related to Google Search please ask them in Google's AMP forum instead of filing an issue here: https://goo.gl/utQ1KZ

If you have a bug or feature request for AMP please fill in the following template.  Delete everything except the headers (including this text).

## What's the issue?

The links provided in the recommendation box of the `amp-next-page` component return a 404 when an AMP article is accessed from the AMP cache.
The links in the recommendation box are rewritten to link to the cached AMP documents. The resultant 404 is due to an incorrect AMP cache URL when the ampUrl is set to a relative URL.

## How do we reproduce the issue?

Sample URL: https://www.bbc.com/mundo/amp/noticias-45657679

1. Go to https://search.google.com/test/amp
2. Paste https://www.bbc.com/mundo/amp/noticias-45657679 in the *Enter a URL to test* field and click the *RUN TEST* button
3. Click *PREVIEW SEARCH RESULT* on the page with the test result.
4. Click the link returned in the search engine result preview.
5. Scroll down to the recommendation box and examine the href value for the link in the `amp-next-page-links` div.
The value is set to
 https://www-bbc-com.cdn.ampproject.org/c/s/www-bbc-com.cdn.ampproject.org/mundo/amp/noticias-43645786
 instead of 
https://www-bbc-com.cdn.ampproject.org/c/s/www.bbc.com/mundo/amp/noticias-43645786
6. Click on the story in the recommendation box to view the `404` response.

download

## What browsers are affected?

All browsers

## Which AMP version is affected?

Version 1809142227020

### PR DESCRIPTION
**Pull request by peterjosling**
Relative page URLs in amp-next-page config don't work right now when served from the cache, as the cache URL already contains a path, which contains the full canonical URL rather than just the document path. Fix by always resolving relative URLs to absolute using the canonical URL first, and then handling it as any other absolute URL (for cache rewriting).

Threw in some added bonus clean up of the `assertConfig` methods :sparkles: 

Fixes #18508

### CODE REVIEW

**Comment by peterjosling** [comment]
@aghassemi

File: extensions/amp-next-page/0.1/config.js [hunk 1]
```diff
@@ -95,13 +90,23 @@ function assertSelectors(selectors) {
...
- * @param {string} origin
- * @param {string=} sourceOrigin
+ * @param {string} documentUrl
  */
-function assertReco(context, reco, origin, sourceOrigin) {
+function assertReco(context, reco, documentUrl) {
+  user().assertString(reco.ampUrl, 'ampUrl must be a string');
+
+  // Rewrite relative URLs to absolute, relative to the canonical URL.
+  const isRelativeUrl = !/^(?:[a-z]+:)?\/\//.test(reco.ampUrl);
```
Comment (aghassemi) [inline extensions/amp-next-page/0.1/config.js]: no need to do the check, it is safe to call `resolveRelativeUrl` regardless. It is no-op if Url is already absolute.

File: extensions/amp-next-page/0.1/config.js [hunk 2]
```diff
@@ -95,13 +90,23 @@ function assertSelectors(selectors) {
...
+  if (isRelativeUrl) {
+    const {canonicalUrl} = Services.documentInfoForDoc(context);
```
Comment (aghassemi) [inline extensions/amp-next-page/0.1/config.js]: the base for resolveRelativeUrl can just be `const base = getSourceUrl(documentUrl);` `canonicalUrl` is the non-AMP version of the AMP page and may actually be on a separate domain than AMP.

File: extensions/amp-next-page/0.1/config.js (hunk 1 above)
Comment (peterjosling) [reply extensions/amp-next-page/0.1/config.js]: Done.

File: extensions/amp-next-page/0.1/config.js (hunk 2 above)
Comment (peterjosling) [reply extensions/amp-next-page/0.1/config.js]: My bad. Fixed.

**Pull request merged**